nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,356 @@
+
+  // If the distance to the goal is less than the motion target distance, i.e.
+  // the 'motion target' is the goal, then we skip the motion target orientation by pointing
+  // it in the same orientation that the last segment of the path
+  double angle_to_target = atan2(motion_target.pose.position.y, motion_target.pose.position.x);
+  if (params_->final_rotation && dist_to_goal < params_->motion_target_dist) {
+    geometry_msgs::msg::PoseStamped stl_pose =
+      transformed_plan.poses[transformed_plan.poses.size() - 2];
+    double dx = goal_pose.pose.position.x - stl_pose.pose.position.x;
+    double dy = goal_pose.pose.position.y - stl_pose.pose.position.y;
+    double yaw = std::atan2(dy, dx);
+    motion_target.pose.orientation.z = sin(yaw / 2.0);
+    motion_target.pose.orientation.w = cos(yaw / 2.0);
+  }
+
+  // Compute velocity command:
+  // 1. Check if we are close enough to the goal to do a final rotation in place
+  // 2. Check if we must do a rotation in place before moving
+  // 3. Calculate the new velocity command using the smooth control law
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header = pose.header;
+  if (params_->final_rotation && (dist_to_goal < goal_dist_tolerance_ || goal_reached_)) {
+    goal_reached_ = true;
+    double angle_to_goal = tf2::getYaw(transformed_plan.poses.back().pose.orientation);
+    cmd_vel.twist = rotateToTarget(angle_to_goal);
+  } else if (params_->initial_rotation && // NOLINT
+    fabs(angle_to_target) > params_->initial_rotation_min_angle)
+  {
+    cmd_vel.twist = rotateToTarget(angle_to_target);
+  } else {
+    cmd_vel.twist = control_law_->calculateRegularVelocity(motion_target.pose);
+  }
+
+  // Transform local frame to global frame to use in collision checking
+  try {
+    costmap_transform_ = tf_buffer_->lookupTransform(
+      costmap_ros_->getGlobalFrameID(), costmap_ros_->getBaseFrameID(),
+      tf2::TimePointZero);
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(
+      logger_, "Could not transform %s to %s: %s",
+      costmap_ros_->getBaseFrameID().c_str(), costmap_ros_->getGlobalFrameID().c_str(),
+      ex.what());
+    return cmd_vel;
+  }
+
+  // Generate and publish local plan for debugging / visualization
+  nav_msgs::msg::Path local_plan;
+  if (!simulateTrajectory(pose, motion_target, local_plan)) {
+    throw nav2_core::NoValidControl("Collision detected in the trajectory");
+  }
+  local_plan.header = transformed_plan.header;
+  local_plan_pub_->publish(local_plan);
+
+  return cmd_vel;
+}
+
+void GracefulMotionController::setPlan(const nav_msgs::msg::Path & path)
+{
+  path_handler_->setPlan(path);
+  goal_reached_ = false;
+}
+
+void GracefulMotionController::setSpeedLimit(
+  const double & speed_limit, const bool & percentage)
+{
+  std::lock_guard<std::mutex> param_lock(param_handler_->getMutex());
+
+  if (percentage) {
+    // Speed limit is expressed in % from maximum speed of robot
+    params_->v_linear_max = std::max(
```

Comment:
What if there are multiple set speed limit updates? this overrides the existing one. If we don't store the original, how can we do a "go 50%" then later "go 75%" of original maximum? This would apply 75% to the 50%'th value